src/main/java/com/google/devtools/build/lib/bazel/commands/ModCommand.java

```diff
@@ -523,6 +563,84 @@ private BlazeCommandResult execInternal(CommandEnvironment env, OptionsParsingRe
     return BlazeCommandResult.success();
   }
 
+  private BlazeCommandResult runTidy(CommandEnvironment env, BazelModTidyValue modTidyValue) {
+    CommandBuilder buildozerCommand =
+        new CommandBuilder()
+            .setWorkingDir(env.getWorkspace())
+            .addArg(modTidyValue.buildozer().getPathString())
+            .addArgs(
+                Stream.concat(
+                        fixupEvents.stream()
+                            .map(RootModuleFileFixupEvent::getBuildozerCommands)
+                            .flatMap(Collection::stream),
+                        Stream.of("format"))
+                    .collect(toImmutableList()))
+            .addArg("MODULE.bazel:all");
+    try {
+      buildozerCommand.build().execute();
+    } catch (InterruptedException | CommandException e) {
+      String suffix = "";
+      if (e instanceof AbnormalTerminationException) {
+        suffix =
+            ":\n" + new String(((AbnormalTerminationException) e).getResult().getStderr(), UTF_8);
+      }
+      return reportAndCreateFailureResult(
+          env,
+          "Unexpected error while running buildozer: " + e.getMessage() + suffix,
+          Code.BUILDOZER_FAILED);
+    }
+
+    for (RootModuleFileFixupEvent fixupEvent : fixupEvents) {
+      env.getReporter().handle(Event.info(fixupEvent.getSuccessMessage()));
+    }
+
+    if (modTidyValue.lockfileMode().equals(LockfileMode.UPDATE)) {
+      // We cannot safely rerun Skyframe evaluation here to pick up the updated module file.
```

Comment:
this part seems rather tricky so I'd just like to flag it so I can think about it more. My immediate reaction was that we should just trigger whatever mechanism we have for detecting diffs/dirtiness (there seem to be quite a few of those... DiffAwareness, DirtinessChecker, etc). But some half-hearted digging revealed that it might be even hackier than what you're doing here. So I'd like to come back later with a clearer head and review this part more carefully.

Reply:
I went with this approach based on @lberki's comments on the design doc. 

Another argument in favor of the current approach is that it could be fully dropped with the proposed new lockfile format: The lockfile wouldn't change at all if only imports are changed.

Reply:
> Another argument in favor of the current approach is that it could be fully dropped with the proposed new lockfile format: The lockfile wouldn't change at all if only imports are changed.

To confirm, this is because we send out the lockfile event (or whatever it's called...) _before_ we check the validity of `use_repo`, right?

Reply:
I'm not sure how your question relates to the quote (did you mean to respond on the thread about Skyframe evaluation failing?). The fixup events are emitted before we check the validity of `use_repo`, which is indeed crucial, as otherwise we would exit the SkyFunction with a failure before we learn of the fixup command. However, this was already necessary before this PR, simply to show the user-facing warning before running into the failure.

Reply:
good, good -- nevermind the quote, not important :)